src/ophyd_async/epics/pvi/pvi.py

```diff
@@ -0,0 +1,323 @@
+    number = match.group(2) or None
+    if number:
+        number = int(number)
+    return name, number
+
+
+def _strip_union(field: Union[Union[T], T]) -> T:
+    if get_origin(field) is Union:
+        args = get_args(field)
+        for arg in args:
+            if arg is not type(None):
+                return arg
+    return field
+
+
+def _strip_device_vector(field: Union[Type[Device]]) -> Tuple[bool, Type[Device]]:
+    if get_origin(field) is DeviceVector:
+        return True, get_args(field)[0]
+    return False, field
+
+
+@dataclass
+class PVIEntry:
+    """
+    A dataclass to represent a single entry in the PVI table.
+    This could either be a signal or a sub-table.
+    """
+
+    name: Optional[str]
+    access: Access
+    values: List[str]
+    # `sub_entries` if the signal is a PVI table
+    # If a sub device is a device vector then it will be represented by a further dict
+    sub_entries: Optional[Dict[str, Union[Dict[int, "PVIEntry"], "PVIEntry"]]] = None
+    device: Optional[Device] = None
+
+    @property
+    def is_pvi_table(self) -> bool:
+        return len(self.values) == 1 and self.values[0].endswith(":PVI")
+
+
+def _verify_common_blocks(entry: PVIEntry, common_device: Type[Device]):
+    if not entry.sub_entries:
+        return
+    common_sub_devices = get_type_hints(common_device)
+    for sub_name, sub_device in common_sub_devices.items():
+        if sub_name in ("_name", "parent"):
+            continue
+        assert entry.sub_entries
+        if sub_name not in entry.sub_entries and get_origin(sub_device) is not Optional:
+            raise RuntimeError(
+                f"sub device `{sub_name}:{type(sub_device)}` was not provided by pvi"
+            )
+        if isinstance(entry.sub_entries[sub_name], dict):
+            for sub_sub_entry in entry.sub_entries[sub_name].values():  # type: ignore
+                _verify_common_blocks(sub_sub_entry, sub_device)  # type: ignore
+        else:
+            _verify_common_blocks(
+                entry.sub_entries[sub_name], sub_device  # type: ignore
+            )
+
+
+_pvi_mapping: Dict[FrozenSet[str], Callable[..., Signal]] = {
+    frozenset({"r", "w"}): lambda dtype, read_pv, write_pv: epics_signal_rw(
+        dtype, "pva://" + read_pv, "pva://" + write_pv
+    ),
+    frozenset({"rw"}): lambda dtype, read_write_pv: epics_signal_rw(
+        dtype, "pva://" + read_write_pv, write_pv="pva://" + read_write_pv
+    ),
+    frozenset({"r"}): lambda dtype, read_pv: epics_signal_r(dtype, "pva://" + read_pv),
+    frozenset({"w"}): lambda dtype, write_pv: epics_signal_w(
+        dtype, "pva://" + write_pv
+    ),
+    frozenset({"x"}): lambda _, write_pv: epics_signal_x("pva://" + write_pv),
+}
+
+
+class PVIParser:
+    def __init__(
+        self,
+        root_pv: str,
+        timeout=DEFAULT_TIMEOUT,
+    ):
+        self.root_entry = PVIEntry(
+            name=None, access=frozenset({"d"}), values=[root_pv], sub_entries={}
+        )
```

Comment:
Why have a root entry rather than a list of child entries?

Reply:
Just to save on code so that `get_pvi_entries` can fetch the top level blocks